### ISSUE TITLE
[Question] Help with ALEBO Model Fits

### ISSUE THREAD

**Issue by wjmaddox** [facebook/Ax#557]
Hi, I'm attempting to use ALEBO to model some high dimensional data and was wondering what the proper way to fit the data using ALEBO should be. I'm mostly focusing on the Hartmann-100 test problem from the original paper right now, and am a bit confused as to if the model fits should be this poor, even with lots of data.

Here's a minimal example describing what's going on with the fits:

```python
import torch

from torch import Tensor

from botorch.test_functions import Hartmann
from botorch.optim.fit import fit_gpytorch_torch
from gpytorch.mlls import ExactMarginalLogLikelihood
from ax.models.torch.alebo import ALEBOGP, ALEBO

class Hartmann6_100(Hartmann):
    def __init__(self, noise_std = 0.0, negate = False):
        super().__init__(dim=6, noise_std=noise_std, negate=negate)
        self._bounds = [(0.0, 1.0) for _ in range(100)]
        self.bounds = Tensor(self._bounds).t()

    def evaluate_true(self, X: Tensor) -> Tensor:
        subset = X[..., [19, 14, 43, 37, 66, 3]]
        return super().evaluate_true(subset)

problem = Hartmann6_100(noise_std=0., negate = True).to(**kwargs)

# use 500 training points
num_init = 500
bounds = problem.bounds
train_x = (bounds[1] - bounds[0]) * torch.rand(num_init, bounds.shape[1], **tkwargs) + bounds[0]
train_y = problem(train_x).unsqueeze(-1)

# construct projection matrix
rank = 12
# i've tried doing the normalization after transposing as well and the model fits are the same
# seems to be what's done in https://github.com/facebook/Ax/blob/eeb23d764fbf2ce5953753bfec745e02b09749f3/ax/modelbridge/strategies/alebo.py#L133
###
projection = torch.randn(train_x.shape[-1], rank, requires_grad=False, **tkwargs)
projection = projection / torch.sqrt((projection ** 2).sum(dim=0))
projection = projection.t()

# prepare an ALEBOGP which has no Laplace approximation
model = ALEBOGP(B = projection, train_X = train_x, train_Y = train_y, train_Yvar = 0.1 * torch.ones_like(train_y))
model.train_inputs = [train_x.matmul(projection.t())] # need to manually do this or it fails

mll = ExactMarginalLogLikelihood(model.likelihood, model)
fit_gpytorch_torch(mll);

# prepare a 500 point test set
test_x = (bounds[1] - bounds[0]) * torch.rand(num_init, bounds.shape[1], **tkwargs) + bounds[0]
test_y = problem(test_x).unsqueeze(-1)

# compute a posterior
model.eval()
pred_post = model.posterior(test_x.matmul(projection.t()))

# model fits are pretty much the mean?
plt.errorbar(
    test_y.cpu(), pred_post.mean.detach().cpu(), yerr=2. * pred_post.variance.detach().cpu() ** 0.5, 
    ls = "none", marker = "x"
)
plt.plot(torch.linspace(0, 3., 50), torch.linspace(0, 3., 50), color = "black")

###### use ALEBO instead?
ax_alebo = ALEBO(projection)

# seems to be the lowest overhead way of doing this?
ax_alebo.model = ax_alebo.get_and_fit_model(
    Xs = [(ax_alebo.B @ X.t()).t() for X in [train_x]], Ys = [train_y], Yvars = [0.1 * torch.ones_like(train_y)]
)
mean, var = ax_alebo.predict(test_x)
# not sure why var is [500, 1, 1,] and not # 500 x 500?

# fits are now just the mean
plt.errorbar(
    test_y.cpu(), mean, yerr=2. * var[...,0] ** 0.5, 
    ls = "none", marker = "x"
)
plt.plot(torch.linspace(0, 3., 50), torch.linspace(0, 3., 50), color = "black")
```

For reference im using gpytorch master, botorch master, and Ax master. along with torch 1.8.0.

**Comment by eytan** [comment]
cc @bletham

**Comment by bletham** [comment]
Hi @wjmaddox, the problem here is that the ALEBO model should only be used on points that were generated inside the embedding.

"Inside the embedding" means that for every `x \in R^D` in the high-dimensional training set, there is a `y \in R^d` such that `x = B^{pinv} y`, where `B^{pinv}` is the pseudo-inverse of the (downwards) projection matrix `B \in R^{d, D}`.

Here you have generated the training set by:
1) Generating points `x_i \in R^D` in the box bounds
2) Evaluating f(x) at those points
3) Projecting the points down to the embedding to get `y_i = B x_i`
4) Fitting a model with training data `(y_i, f(x_i))`.

This is bad because there is significant projection error being introduced into the embedding. "Projection error" means the phenomenon that there may be two points `x_i` and `z_i` that project down to the same `y_i`, but have very different values for `f(x_i)` and `f(z_i)`. That is, `y_i = B x_i = B z_i` but `f(x_i) != f(z_i)`. This is obviously bad for GPs because we will have conflicting data at `y_i` that violates the smoothness assumption we are making by trying to fit a GP in the embedding.  This paper has a nice description of the issue of projection error with some visualizations: https://link.springer.com/article/10.1007/s11590-019-01433-w

In that paper, projection error is handled with quantile GP regression. In REMBO and its variants (including ALEBO), it is handled by keeping the projection fixed and only evaluating data points inside the embedding. This way there is never any projection error (there is a one-to-one mapping between the embedding and the high-dimensional space for such points) and we can be confident that if a GP fits well in the true subspace, it will fit well in the embedding (subject to the kernel details as discussed in the ALEBO paper).

So, the correction to generating the training set would then be:
1) Generate points in the embedding `y_i \in R^d`
2) Project them up to get the (unique) high-dimensional point corresponding to y_i that is in the embedding: `x_i = B^{pinv} y_i`
3) Fitting a model with training data `(y_i, f(x_i))`.

Care must be taken in step (2) to ensure that `x_i` satisfies high-dimensional box bounds. This can be done by applying the constraints described in the paper when generating points in the embedding in (1).

The same considerations apply to the random initialization typically used in Bayesian Optimization; it cannot be any space-filling design in the high-dimensional space, it must inside the embedding. Our code for ALEBO includes a function for generating an appropriate space-filling design in the high-dimensional space that satisfies the "in the embedding" property of `x = B^{pinv} B x`. You can use that function to generate appropriate train/test sets like we did here in our code for Fig. 3 of the paper:
https://github.com/facebookresearch/alebo/blob/2f8c7bafb3611c1e91bfed4a5a4e6690ef12bdbb/figs/fig_3.py#L40-L43

Generating the train/test set that way, you will see good model performance.

A few additional comments:

1) The `train_x` input to `ALEBOGP` should be the projected, low-dimensional points, not the high-dimensional X, which is why you're seeing a need to project the `train_inputs`. You can see that in our code the projection happens here, prior to constructing the GP: https://github.com/facebook/Ax/blob/eeb23d764fbf2ce5953753bfec745e02b09749f3/ax/models/torch/alebo.py#L588-L593

2) The reason the output of `ALEBO.predict` is [500, 1, 1] and not [500, 500] is because in Ax the API allows for specifying covariance across outputs (of which there is 1 here) rather than covariance across predict points: https://github.com/facebook/Ax/blob/master/ax/models/torch_base.py#L53-L64

Let me know if you have other questions!

**Comment by lena-kashtelyan** [comment]
@wjmaddox, hopefully @bletham's answer above covered your question –– please feel free to reopen the issue if you have follow-ups!

**Comment by wjmaddox** [comment]
Sorry, this got de-prioritized recently. Thanks for the help @bletham, I was able to get very sensible fits in the alebo subspace as a result. Still trying to run a proper BO loop (the subspace doesn't seem to contain an optimum), but I'll re-open if/when I get a chance to look at it again.
